ddtrace/internal/core.py

```diff
@@ -142,7 +144,7 @@ def has_listeners(self, event_id):
     def on(self, event_id, callback):
         # type: (str, Callable) -> None
         if callback not in self._listeners[event_id]:
-            self._listeners[event_id].append(callback)
+            self._listeners[event_id].insert(0, callback)
```

Comment:
From the perspective of `internal.core`, the order here is arbitrary and there's no inherent reason to add listeners to the beginning or end of the list. The reason I made this change here is related to the order in which `on` is called by the AppSec and Tracing products.

In general, AppSec callbacks tend to depend on objects that Tracing callbacks created. During library startup, the AppSec `on` calls happen before those from Tracing. If `on` was `append`ing to the callback list, this would result in the AppSec listener being in the list before the Tracing one it depends on. This change swaps that order for the default common case.

In the process of writing this change, I realized that temporal dependencies between listeners of the same event are confusing and lead to difficult bugs, so in many such cases I've split multi-listener events into multiple differently-named events, allowing us more explicit control over the relative ordering and argument lists.

Reply:
Got it, yeah ok this is what I thought was going on but wasn't sure. I also then got worried about how the order of listeners could affect the result. The idea of splitting multi-listener events is helpful. Will try to keep that in mind.